sql-tuning-best-practice.md

```diff
@@ -0,0 +1,954 @@
+
+- Large-Scale Data Analysis: For queries involving extensive data scans, such as OLAP workloads, TiFlash can deliver faster performance compared to TiKV due to its columnar storage and MPP capability.
+- Complex Scans, Aggregations and Joins: TiFlash is ideal for queries with heavy aggregations and joins, as it can process these operations more efficiently by only reading necessary columns.
+- Mixed Workloads: In hybrid environments where both transactional (OLTP) and analytical (OLAP) workloads run simultaneously, TiFlash can handle analytical queries without impacting TiKV’s performance for transactional queries.
+- SaaS Arbitrary Filtering Workloads: In SaaS workloads, there is often a need for arbitrary filtering across many columns. Indexing all columns is impractical, especially when the primary key includes a tenant ID and all queries include this ID. TiFlash is an ideal solution here: because data in TiFlash is sorted and clustered by primary key, it enables efficient table range scans. This allows TiFlash to provide fast performance without the overhead of maintaining multiple indexes.
+
+Using TiFlash strategically can enhance query performance and optimize resource usage in TiDB for data-intensive, analytical queries. Here are two use case of TiFlash
+
+### Analytical Query
+
+The original query requires joining the order_line and item tables. The plan for executing this query on the TiKV storage engine takes 21.1 seconds.
+
+In the TiKV plan:
+
+- TiDB needs to fetch 3,864,397 rows from the lineitem table and 10 million rows from the part table.
+- The hash join operation (HashJoin_21) is performed at the TiDB layer, along with the subsequent projection (Projection_38) and aggregation (HashAgg_9) operations.
+
+The plan for executing the same query on the TiFlash MPP (Massively Parallel Processing) takes only 1.41 seconds, which is 15 times faster than the TiKV plan.
+
+In the TiFlash MPP plan:
+
+- The optimizer recognizes that both the order_line and item tables have TiFlash replicas available.
+- As a result, the optimizer is able to push down the entire query execution, including the table scans, hash join, column projection, and aggregation, to the TiFlash MPP layer.
+- Performing the entire query execution on the TiFlash MPP layer, which is optimized for analytical workloads, leads to the significant performance improvement compared to the TiKV plan.
+
+The key difference between the two plans is that the TiFlash MPP plan can leverage the distributed processing capabilities of the TiFlash MPP layer to execute the entire query more efficiently, while the TiKV plan requires more coordination between the TiDB and TiKV components.
+
+```sql
+select 100.00 * sum(case when i_data like 'PR%' then ol_amount else 0 end) / (1+sum(ol_amount)) as promo_revenue
+from	order_line, item
+where	ol_i_id = i_id and ol_delivery_d >= '2007-01-02 00:00:00.000000' and ol_delivery_d < '2030-01-02 00:00:00.000000';
```

Comment:
Are the columns really named like this? Like `ol_i_id` for `order_line`.`id`, and `i_id` for `item`.`id`?